task done

MikhailKamensky opened the pull request:
Добрый день. Благодарю за сохраненную доработку) Создал новый проект и новый репозиторий, чтобы корректно отражались все изменения

Commit:
task done

okryukova commented on pom.xml:
⚠️Можно улучшить. Хорошей практикой является добавление README.md с описанием проекта, которое включает название, технологии в проекте (например, Java 11, JUnit 4.13.2, maven 3.9.0 и т.д.), настройка (если необходима) и как запускать (mvn clean test - кавычки позволяют выполнять запуск из файла README.md). Для оформления данных в виде таблице используй сочетание клавиш Ctrl + N или Cmd + N. Как оформлять README.md можно ознакомиться в статье https://blog.jetbrains.com/idea/2022/08/markdown-support-improvements/

MikhailKamensky replied:
Сделано

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/FirstOrderForm.java:
```diff
@@ -0,0 +1,62 @@
+
+    //локаторы
+
+    private By firstOrderPage = By.xpath(".//div[@class='Order_Content__bmtHS']");
```
✅Отлично. Молодец, что используешь относительные , а не полные локаторы. Такой код меньше подвержен изменениям.

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/FirstOrderForm.java:
```diff
@@ -0,0 +1,62 @@
+
+
+    public void selectMetroStation (String stationName) {
+        driver.findElement(By.xpath(".//div[text()='"+stationName+"']")).click();
```
⛔️Нужно исправить. Локаторы должны хранится в полях класса. Поправь, пожалуйста, во  всем коде

MikhailKamensky replied:
Сделано

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/ScooterHomePage.java:
```diff
@@ -0,0 +1,128 @@
+    public final String answerText7 = "Да, обязательно. Всем самокатов! И Москве, и Московской области.";
+
+
+    private By accordionPanel0 = By.xpath(".//div[@class='accordion']/div[1]");
```
⛔️Нужно исправить. accordionPanel0 -- неудачный нейминг. Что такое accordion? Плохая идея давать имя переменной исходя из класса. Этот класс может быть задан и для других элементов. Также при нейминге не рекомендуется использовать числа (Field2), их еще называют magicNumbers. Очень тяжело поддерживать код с magicNumbers. Лучше привязываться к теме вопроса.Подбери, пожалуйста, более удачные названия для всех переменных accordionPanel и accordionText (можно использовать слова question и answer).

MikhailKamensky replied:
Сделано

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/ScooterHomePage.java:
```diff
@@ -0,0 +1,128 @@
+    private By accordionPanel1 = By.xpath(".//div[@class='accordion']/div[2]");
+    private By accordionPanel2 = By.xpath(".//div[@class='accordion']/div[3]");
+    private By accordionPanel3 = By.xpath(".//div[@class='accordion']/div[4]");
+    private By accordionPanel4 = By.xpath(".//div[@class='accordion']/div[5]");
```
⚠️Можно улучшить. Локаторы отличаются только индексом строки. Подумай, как можно уменьшить кол-во локаторов

MikhailKamensky replied:
сделано

MikhailKamensky replied:
Сделано (шаблон локаторы + метод, возвращающий локатор)

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/ScooterHomePage.java:
```diff
@@ -0,0 +1,128 @@
+        driver.findElement(accordionPanel5).click();
+    }
+
+    public void clickAccordionPanel6 () {
```
⛔️Нужно исправить.  Неинформативные названия методов

okryukova commented on src/test/java/OrderButtonsTest.java:
```diff
@@ -0,0 +1,59 @@
+    @Test
+    public void checkSmallOrderButton() {
+
+        driver.get("https://qa-scooter.praktikum-services.ru/");
```
⚠️Можно улучшить. Url "https://qa-scooter.praktikum-services.ru/" обычно выделяют как константу (можно реализовать в классе MainPage).

okryukova commented on src/test/java/OrderButtonsTest.java:
```diff
@@ -0,0 +1,59 @@
+        FirstOrderForm  objFirstOrderPage = new FirstOrderForm(driver);
+
+        objHomePage.clickSmallOrderButton();
+        WebElement element = driver.findElement(By.className("Order_Content__bmtHS"));
```
⛔️Нужно исправить. Не следует описывать локаторы внутри тестовых методов. Используй инкапсуляцию. В классах page objects опиши все веб-элементы, которые используются в кейсах, с помощью локаторов и укажи модификатор доступа private. Реализуй шаги для действий над веб-элементами. Тесты должны состоять из шагов (для получения актуального результата) и проверки (сравнение ожидаемого и актуального результата). Это позволит переиспользовать веб-элементы в разных шагах(методах), а шаги в разных тестах.

okryukova commented on src/test/java/OrderButtonsTest.java:
```diff
@@ -0,0 +1,59 @@
+    @Before
+    public void startUp() {
+        WebDriverManager.chromedriver().setup();
+        driver = new ChromeDriver();
```
⚠️Можно улучшить. Браузер, как правило, запускают в полноэкранном режиме (maximized) или определенном размере окна.

MikhailKamensky replied:
Сделано

okryukova commented on src/test/java/OrderButtonsTest.java:
```diff
@@ -0,0 +1,59 @@
+        WebElement element = driver.findElement(By.className("Order_Content__bmtHS"));
+    }
+
+    @After
```
⚠️Можно улучшить. Блоки открытия и закрытия браузера одинаковый для всех тестовых классов. Целесообразно вынести его в базовый или утилитный класс.

okryukova commented on src/test/java/OrderFormTest.java:
```diff
@@ -0,0 +1,104 @@
+        SecondOrderForm objSecondOrderPage = new SecondOrderForm(driver);
+
+        objFirstOrderPage.enterFirstName(firstName);
+        objFirstOrderPage.enterLastName(lasnName);
```
⚠️Можно улучшить. В теме UI-тестирование > Page Object > Работа с методами > Объединение методов хорошо описан прием объединения методов. Применяй его для сокращения кода автотеста, где это необходимо. Например, при заполнении форм.

MikhailKamensky replied:
Сделано

okryukova commented on src/test/java/QuestionsHomeTests.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    @Test
+    public void checkAccordionPanel0() {
```
⛔️Нужно исправить. Все тесты класса аналогичны, выполняют одни и те же действия. Есть отличие только в номере строки и тексте проверяемого ответа. Подумай, как можно сократить кол-во кода. Попробуй использовать параметризацию.

MikhailKamensky replied:
Сделано

okryukova commented on src/test/java/OrderButtonsTest.java:
```diff
@@ -0,0 +1,59 @@
+
+
+        objHomePage.clickSmallOrderButton();
+        WebElement element = driver.findElement(By.className("Order_Content__bmtHS"));
```
⛔️Нужно исправить.  В каждом тесте должен быть ассерт. Добавь, пожалуйста

MikhailKamensky replied:
Сделано

Commit:
added README.MD
locators moved to class fields
renamed former "accordionPanelX" variables and locators
added maximized browser window
added asserts to all tests

MikhailKamensky commented:
Отработал не все замечания, но время на 1 доработку истекает, оставляю их на вторую

MikhailKamensky commented:
Не уверен, что правильно отправляю задание на проверку - сделал по инструкции: правки, коммит, пуш, ссылка на этот же мердж

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/ScooterHomePage.java:
```diff
@@ -0,0 +1,128 @@
+    private WebDriver driver;
+
+    public final String answerPrice = "Сутки — 400 рублей. Оплата курьеру — наличными или картой.";
+    public final String answerFewScootersRent = "Пока что у нас так: один заказ — один самокат. Если хотите покататься с друзьями, можете просто сделать несколько заказов — один за другим.";
```
⚠️Можно улучшить.  Советую добавить ключевое слово  static. Тогда это будут константы. К ним можно будет ображаться по имени класса

MikhailKamensky replied:
Сделано

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/ScooterHomePage.java:
```diff
@@ -0,0 +1,128 @@
+        driver.findElement(TodayRentPanel).click();
+    }
+
+    public void clickAccordionPanel4 () {
```
⛔️Нужно исправить. неудачный нейминг. При нейминге не рекомендуется использовать числа (Panel4), их еще называют magicNumbers. Очень тяжело поддерживать код с magicNumbers. Лучше привязываться к теме вопроса.Подбери, пожалуйста, более удачные названия для всех методов

MikhailKamensky replied:
Сделано (однотипные методы заменены на один с параметром)

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/ScooterHomePage.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    public String getAccordionPanel4Text () {
+        return driver.findElement(answerRentExtensionText).getText();
```
⚠️Можно улучшить. 8 повторяющихся методов на получение текста ответа. Подумай, как можно их объеденить. Может быть стоит использовать индекс вопроса? Его можно передавать как параметр метода.

MikhailKamensky replied:
Сделано

okryukova commented on src/test/java/OrderFormTest.java:
```diff
@@ -0,0 +1,105 @@
+        objFirstOrderPage.enterAdress(adress);
+        objFirstOrderPage.setMetroStation();
+
+        WebElement element = driver.findElement(By.xpath(".//div[text()='" + station + "']"));
```
⛔️Нужно исправить. Локаторы в тестах на прямую не используем. Локатор нужно перенести в РО и написать метод работы с ним. Поправь, пожалуйста, этот момент во всем коде

MikhailKamensky replied:
Сделано

okryukova commented on src/test/java/QuestionsHomeTests.java:
```diff
@@ -0,0 +1,213 @@
+        ScooterHomePage objHomePage = new ScooterHomePage(driver);
+
+
+        WebElement element = driver.findElement(By.className("accordion"));
```
⛔️Нужно исправить.  Использование параметризации в FQA тестах обязательно для зачета. Переделай, пожалуйста, тесты

MikhailKamensky replied:
Сделано

Commit:
тексты ответов объявлены как константы; изменен нейминг методов в ScooterHomePage.java (убраны "магические числа"); в ScooterHomePage.java повторяющиеся локаторы заменены на связку "шаблон+метод, возвращающий локатор исходя из входящего параметра"; в ScooterHomePage.java однотипные методы заменены на единый метод с входящим параметром; в OrderFormTest тестовые операции объединены в смысловые шаги; в QuestionsHomeTests однотипные тесты объединены в один параметризованный тест; все оставшиеся в тестовых файлах локаторы вынесены в POM-файлы.

okryukova commented on src/main/java/ru/praktikum_serveces/qa_scooter/ScooterHomePage.java:
```diff
@@ -0,0 +1,176 @@
+//    private By answerPriceText = By.xpath(".//p[text() = '"+ answerPrice +"']");
+//    private By answerFewScootersRentText = By.xpath(".//p[text() = '"+ answerFewScootersRent +"']");
+//    private By answerRentTimeText = By.xpath(".//p[text() = '" + answerRentTime + "']");
+//    private By TodayRentText = By.xpath(".//p[text() = '"+ answerTodayRent +"']");
```
⛔️Нужно исправить. Не используемый код нужно удалить. Нужно поправить этот момент во всем проекте.

MikhailKamensky replied:
Сделано

okryukova commented on src/test/java/QuestionsHomeTests.java:
```diff
@@ -0,0 +1,256 @@
+
+        objHomePage.clickQuestionPanel(questionNum);
+        new WebDriverWait(driver, Duration.ofSeconds(3))
+                .until(ExpectedConditions.visibilityOfElementLocated(objHomePage.createAnswerPanelLocator(answer)));
```
⛔️Нужно исправить. В тестах не должно быть явного кода работы с WebDriver и WebDriverWait. В классах  Page  Object описываем локаторы поиска элементов и методы для работы с элементами, в тестах используем готовые методы.

MikhailKamensky replied:
Сделано

Commit:
удален неиспользуемый код; код, работающий с WebDriver/WebDriverWait вынесен в pom-файлы; исправлена ошибка падения тестов формы заказа на этапе выбора станции (добавлен метод прокрутки к требуемому элементу)
